clippy_lints/src/useless_conversion.rs

```diff
@@ -81,9 +81,16 @@ impl<'tcx> LateLintPass<'tcx> for UselessConversion {
                     }
                 }
                 if is_trait_method(cx, e, sym::IntoIterator) && name.ident.name == sym::into_iter {
-                    if let Some(parent_expr) = get_parent_expr(cx, e) {
-                        if let ExprKind::MethodCall(parent_name, ..) = parent_expr.kind {
-                            if parent_name.ident.name != sym::into_iter {
+                    if matches!(recv.kind, ExprKind::Path(..)) && get_parent_expr(cx, e).is_some() {
+                        if_chain! {
```

Comment:
I don't quite get what this new condition is trying to check for. There also aren't any new tests added that this new code should block the linting for. Can you explain this a bit, please?

Reply:
Sure. As explained in the commit message, two problems can occur by removing `.into_iter()` even when it is the identity function:

- If the receiver of `.into_iter()` is an immutable local variable `x`, removing `.into_iter()` in the context of a larger expression can change the mutability. For example, `x.into_iter().next()` works, as `x.into_iter()` won't. Note that if we are not in the context of a larger expression, we can remove the call: `let mut y = x.into_iter();` and `let mut y = x;` are equivalent, same without `mut`.
- If the receiver is a `const`, Rust can start displaying warnings although the transformation is correct. For example, if `R` is a `const` `Range<_>` (which implements `Iterator`), `R.into_iter().next()` doesn't trigger any warning while `R.next()` warns that `R` starts from scratch at every call to `.next()` (which is also the call of `R.into_iter().next()`. Here also this can only happen in the context of a larger expression, `let y = R.into_iter();` and `let y = R;` do the same thing and trigger no warning.

To avoid doing an incorrect transformation (first case) or introducing warnings from the compiler (second case), we skip attempting to removing `.into_iter()` if:

- The receiver is a `ExprKind::Path` (to cover local variables and const items)
- and the call to `.into_iter()` is used in the context of a larger expression

In this case, we want to skip the `.into_iter()` removal unless we can check (`if_chain!` below) that the receiver is a mutable local variable, because there we have no problem removing the `.into_iter()`.

One of the added tests checks the block, in addition to the existing `test_issue_5833()` which already covers the case of the receiver being a immutable local variable.

This one 

```rust
fn test_const_into_iter1() -> Result<(), ()> {
    const NUMBERS: std::ops::Range<i32> = 0..10;
    let _ = NUMBERS.into_iter().next();

    Ok(())
}
```

checks that it is blocked as well (`const` item `NUMBERS`, which is a `ExprKind::Path` and not a mutable local variable). We do not want to remove `.into_iter()` there (although it would be correct) because of the compiler warning it would provoke.

For info, I started this yesterday when I was helping a friend of mine with their Rust code for advent-of-code. They used `s.lines().into_iter().map(…)` and when I told them that `.into_iter()` was unnecessary as `.lines()` returned an `impl Iterator` they told me that Clippy hadn't warned them.

Reply:
Thanks for the great explanation! And sorry for missing the commit message.

Usually we don't care about a follow up warning by rustc (or Clippy), so you don't really have to handle that case. In the range example, I would even argue that suggesting to remove `into_iter` and then get a follow up warning is beneficial, because it points out an additional potential flaw in the code.

Reply:
Makes sense. I will push an amended patch. Also, I've identified another problematic corner case that I will integrate: if the receiver `x` implements `Copy`, `x.into_iter().next()` will not advance `x` while `x.next()` will (I identified this while thinking about `static` variables which could be used as receiver if they implement `Copy`). I will include new tests for this as well.